src/main/java/com/snowflake/kafka/connector/internal/streaming/SchematizationUtils.java

```diff
@@ -0,0 +1,223 @@
+import static org.apache.kafka.connect.data.Schema.Type.INT64;
+import static org.apache.kafka.connect.data.Schema.Type.STRING;
+import static org.apache.kafka.connect.data.Schema.Type.STRUCT;
+
+import com.snowflake.kafka.connector.internal.SnowflakeConnectionService;
+import com.snowflake.kafka.connector.internal.SnowflakeErrors;
+import com.snowflake.kafka.connector.internal.SnowflakeKafkaConnectorException;
+import com.snowflake.kafka.connector.records.RecordService;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import net.snowflake.client.jdbc.internal.fasterxml.jackson.databind.JsonNode;
+import org.apache.kafka.connect.data.Field;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.Schema.Type;
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** This is a class containing the helper functions related to schematization */
+public class SchematizationUtils {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SchematizationUtils.class);
+
+  /**
+   * Transform the objectName to uppercase unless it is enclosed in double quotes
+   *
+   * <p>In that case, drop the quotes and leave it as it is.
+   *
+   * @param objectName name of the snowflake object, could be tableName, columnName, roleName, etc.
+   * @return Transformed objectName
+   */
+  public static String formatName(String objectName) {
+    return (objectName.charAt(0) == '"' && objectName.charAt(objectName.length() - 1) == '"')
+        ? objectName.substring(1, objectName.length() - 1)
+        : objectName.toUpperCase();
+  }
+
+  /**
+   * Execute a ALTER TABLE command if there is any extra column that needs to be added, or any
+   * column nullability that needs to be updated, used by schema evolution
+   *
+   * @param conn connection to the Snowflake
+   * @param tableName table name
+   * @param nonNullableColumns a list of columns that needs to update the nullability
+   * @param extraColNames a list of columns that needs to be updated
+   * @param record the sink record that contains the schema and actual data
+   */
+  public static void evolvingSchemaIfNeeded(
+      SnowflakeConnectionService conn,
+      String tableName,
+      List<String> nonNullableColumns,
+      List<String> extraColNames,
+      SinkRecord record) {
+    // Update nullability if needed, ignore any exceptions since other task might be succeeded
+    if (nonNullableColumns != null) {
+      try {
+        conn.alterNonNullableColumns(tableName, nonNullableColumns);
+      } catch (SnowflakeKafkaConnectorException e) {
+        LOGGER.warn(
+            String.format(
+                "Failure altering table to update nullability: %s, this could happen when multiple"
+                    + " partitions try to alter the table at the same time and the warning could be"
+                    + " ignored",
+                tableName),
```

Comment:
should we instead prevent this from happening?

Reply:
makes sense. i was thinking in terms of whether this will run into an loop but eventually it will stop once we change the schema.